bitcoin/src/lib.rs

```diff
@@ -42,12 +42,14 @@
 #![allow(clippy::manual_range_contains)] // More readable than clippy's format.
 #![allow(clippy::needless_borrows_for_generic_args)] // https://github.com/rust-lang/rust-clippy/issues/12454
 
-// Disable 16-bit support at least for now as we can't guarantee it yet.
-#[cfg(target_pointer_width = "16")]
-compile_error!(
-    "rust-bitcoin currently only supports architectures with pointers wider than 16 bits, let us
-    know if you want 16-bit support. Note that we do NOT guarantee that we will implement it!"
```

Comment:
I've had another think and I still can't work it out - _why_ cant' we guarantee that `rust-bitcoin` works on 16-bit targets? Some ideas
- because we assume 32 bits and do a bunch of casts?
- because hashes would slow as hell if `u32` was two words?
- because of some crypto dark-magic that I don't understand?

Reply:
Because we are using `usize` for the sizes of containers, which (by consensus) should be allowed to have more than 65536 entries in them.

Reply:
> * because we assume 32 bits and do a bunch of casts?

Yes, this one. And @apoelstra is correct too though there's not much choice in 16-bit architectures - you can't bypass the limitation by just using `u32`, because the hardware simply cannot address that.

What could work is allowing the small crates that are unaffected by consensus rules to compile on 16-bits. `units`, `addresses`, `hashes`, ... It could be useful for the DIY embedded enthusiasts. Use cases include making calculators, block clocks, POS devices... It's mostly toys though I don't think it'd be too much work if we restrict ourselves to the obvious crates.

Reply:
Maybe *only* `primitives` should block it because that's where the consensus stuff is (and where the majority of the casts are, I'll bet).

Elsewhere we can audit the casts. I'll bet there aren't many cases outside of consensus parsing where we're casting numbers that might be big enough to overflow u16, but are guaranteed not to overflow u32.

Reply:
Good point! Though crates depending on `primitives` will get the restriction transitively.

Reply:
True, but I think that's "working as intended".

Reply:
Indeed.